tests/beta/test_default_qubit_tf.py

```diff
@@ -1045,174 +1020,195 @@ def circuit(a, b):
 
-        return cost_fn
-
-    @pytest.mark.parametrize("interface", ["tf"])
-    def test_tf_interface_device_diff(self, cost, interface, tol):
-        """Tests that the gradient of an arbitrary U3 gate is correct
-        using the TensorFlow interface"""
-        import tensorflow as tf
+        theta = 0.543
+        phi = -0.234
+        lam = 0.654
 
-        params = tf.Variable(self.p, dtype=tf.float64)
+        params = tf.Variable([theta, phi, lam], dtype=tf.float64)
 
         with tf.GradientTape() as tape:
             tape.watch(params)
             res = cost(params)
 
-        assert np.allclose(res.numpy(), self.expected_cost, atol=tol, rtol=0)
+        # check that the result is correct
+        expected_cost = (np.sin(lam) * np.sin(phi) - np.cos(theta) * np.cos(lam) * np.cos(phi)) ** 2
+        assert np.allclose(res.numpy(), expected_cost, atol=tol, rtol=0)
 
         res = tape.gradient(res, params)
-        assert np.allclose(res.numpy(), self.expected_grad, atol=tol, rtol=0)
 
-    @pytest.fixture
-    def cost_with_decomposition(self, interface, torch_support):
-        """Fixture to create cost function for the test class"""
+        # check that the gradient is correct
+        expected_grad = (
+            np.array(
+                [
+                    np.sin(theta) * np.cos(lam) * np.cos(phi),
+                    np.cos(theta) * np.cos(lam) * np.sin(phi) + np.sin(lam) * np.cos(phi),
+                    np.cos(theta) * np.sin(lam) * np.cos(phi) + np.cos(lam) * np.sin(phi),
+                ]
+            )
+            * 2
+            * (np.sin(lam) * np.sin(phi) - np.cos(theta) * np.cos(lam) * np.cos(phi))
+        )
+        assert np.allclose(res.numpy(), expected_grad, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("interface", ["autograd", "torch"])
+    def test_error_backprop_wrong_interface(self, interface, tol):
+        """Tests that an error is raised if diff_method='backprop' but not using
+        the TF interface"""
         dev = qml.device("default.qubit.tf", wires=1)
 
-        @qml.qnode(dev, diff_method="backprop", interface=interface)
-        def circuit(x, weights, w=None):
-            """In this example, a mixture of scalar
-            arguments, array arguments, and keyword arguments are used."""
-            qml.QubitStateVector(1j * np.array([1, -1]) / np.sqrt(2), wires=w)
-            # the parameterized gate is one that gets decomposed
-            # via a template
-            qml.U3.decomposition(x, weights[0], weights[1], wires=w)  # <--- decomposition is used
+        def circuit(x, w=None):
+            qml.RZ(x, wires=w)
             return qml.expval(qml.PauliX(w))
 
-        def cost_fn(params):
-            """Perform some classical processing"""
-            return circuit(params[0], params[1:], w=0) ** 2
+        with pytest.raises(
+            ValueError,
+            match="default.qubit.tf only supports the tf interface when diff_method='backprop'",

```

Comment:
This error message seems to be the reverse of what we want. We should allow tf interface for any diff method, but we can't use diff_method="backprop" unless it's the tf interface

Reply:
Good catch, don't know how I got it backwards

Reply:
Looks like it was already in the codebase!

Reply:
Fixed